### ISSUE TITLE
Type variable scoping with variadic generics

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#4949]
### Discussed in https://github.com/microsoft/pyright/discussions/4918

<div type='discussions-op-text'>

<sup>Originally posted by **sylee957** April  6, 2023</sup>
I also found an issue with #4852 with variadic generics

Given the functions

```python3
from collections.abc import Callable
from typing import TypeVar, TypeVarTuple

X_0 = TypeVar("X_0")
X_1 = TypeVar("X_1")
Xs = TypeVarTuple("Xs")
Ys = TypeVarTuple("Ys")
X = TypeVar("X")
Y = TypeVar("Y")


def parallel(
    f: Callable[[X], Y],
    g: Callable[[*Xs], tuple[*Ys]],
) -> Callable[[X, *Xs], tuple[Y, *Ys]]:
    def wrapped(a: X, *bs: *Xs):
        return f(a), *g(*bs)

    return wrapped


def identity(x: X) -> X:
    return x


def parallel_identity(*xs: *Xs) -> tuple[*Xs]:
    return xs
```

This does not pass the type check

```python3
def test_1(a: X_0, b: X_1, *others: *Xs) -> tuple[X_0, X_1, *Xs]:
    return parallel(identity, parallel(identity, parallel_identity))(a, b, *others)
```

While the same trick with duplicate functions works

```python3
def identity_1(x: X) -> X:
    return x


def identity_2(x: X) -> X:
    return x


def test_2(a: X_0, b: X_1, *others: *Xs) -> tuple[X_0, X_1, *Xs]:
    return parallel(identity_1, parallel(identity_2, parallel_identity))(a, b, *others)
```</div>

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.312, which I just published. It will also be included in the next insiders release of pylance.
